Improvements to Apple Pay\Google Pay when handling subscription variations

mattallan opened the pull request:
Fixes #2755 

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->


Recently we merged https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2736 which fixed an issue with synced subscriptions that required shipping not being supported by the Payment Request Buttons (PRB). We patched this by simply not loading the PRB's for synced subscriptions that didn't have an initial payment and required shipping.

While reviewing the same changes made to WooPayments (https://github.com/Automattic/woocommerce-payments/pull/7582) we discovered we can make improvements to how this function handles variable subscriptions. Previously we would completely hide the PRB's if **any** of the variations didn't support PRB, but turns out we can tweak the new `product_has_trial_and_needs_shipping()` function so that we only hide the PRB is _all_ variations, then when a variation is selected, check whether that specific variation supports PRB and hide/show the buttons accordingly:

![Screen Capture on 2023-11-03 at 12-48-54](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/3ddad22d-d0cb-43da-a45a-5f3259d4cfab)

While bringing over the same changes I made to WooPayments, I found a couple more issues with Stripe's payment request button implementation. These are all getting fixed by this PR:
1. `ajax_get_selected_product_data()` wasn't building the correct product data for subscription variations. This was because we were only looking for `'variable' === $product->get_type()` and not checking for `'variable-subscription'`
2. Exceptions thrown by the `ajax_get_selected_product_data()` function weren't hiding the PRB on the front end.
3. Virtual synced subscription products display the wrong total when purchased from the product page. This is because the `get_product_price()` function cannot accurately calculate the prorated sync price or incorporate mock trials. This only happens during `cart->calculate_totals()`
4. If you have selected a variation that is not supported by PRB and you click the add to cart button, the Apple Pay and Google Pay buttons were displayed on the page after adding the product to the cart.

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

#### Testing Synced Variations

1. Enable Apple Pay and Google Pay payment methods
2. Install and active Woo Subscriptions
5. Enable synchronised subscriptions in **WooCommerce > Settings > Subscriptions > Synchronise renewals**
![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/336e0a3d-c6c9-4978-a66a-9eb0d1d28628)
6. Create a variable subscription product where one is synced and virtual and the other is synced and not virtual and is not synced to today.
7. While on the `develop` branch, view the product page and note the Payment Request buttons (PRB) aren't loaded.
8. Check out this branch and confirm the PRB are loaded on product pages and only hidden after selecting the variation that is synced and requires shipping.

#### Incorrect ajax_get_selected_product_data()

1. Create a variable subscription with at least 2 variations that cost different amounts
2. While on `develop`, view the product page and select one variation and click on the GPay button
3. Close the GPay window and choose the other variation and click on the GPay button again
4. Notice the price doesn't change?

https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/c19229e7-acaa-416b-bb02-1d104740549d

When you run the same test on this branch, the correct prices are sent to the Payment Request Buttons.

> [!NOTE]
> I'm noticing the shipping values aren't being added correctly to the total but this isn't new to this PR. I'll keep exploring this issue to see if there's a simple solution.


---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

Commit:
Refactor product_has_trial_and_needs_shipping() to return true if all variations pass instead of at least one variation

Commit:
When a variation is selected, run product_has_trial_and_needs_shipping() on the variation

Commit:
Fetch the correct product data for subscription variations

Commit:
If an exception is thrown during getSelectedProductData(), hide the payment request buttons

mattallan commented:
> I'm noticing the shipping values aren't being added correctly to the total but this isn't new to this PR. I'll keep exploring this issue to see if there's a simple solution.

I've pushed a fix up in a separate PR since this was also impacting regular WC variable products and not just subscription variations: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2748

Commit:
Update get_product_data() to check for variable subscription products as well

Commit:
is_payment_upfront() already checks if the first payment is today

mattallan committed:
Merge branch 'develop' into 2395-further-improvements

mattallan commented:
Just making a comment that no new changelog is needed as the one we added in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2736 is relevant for this PR.

Mayisha reviewed (commented):
Thanks @mattallan 

The PRB on variable subscription page is working perfectly as described. It shows up for the digital product with future sync and physical product without future sync :+1: 

Just noticed one small thing. When the default variation is not set (`choose an option` being the selected option), the PRB appears enabled on screen.

![gpay](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/96ed41b7-c915-4f7a-ac9a-46ce8ca23e48)

I got another issue which is probably related to the `Incorrect ajax_get_selected_product_data()` fix mentioned in the description. I have created a variable product with the following variations-
- Physical product without sync (price 20)
- Physical product with sync on 4th day (price 10)
- Virtual product with sync on 4th day (price 100)

When I click GPay for the virtual product the price is off.
![Peek 2023-11-10 12-06](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/38a62be9-96cc-4237-ac9e-c4ac5b88c287)

mattallan committed:
Merge branch 'develop' into 2395-further-improvements

Commit:
Fix whitespace issues caused by resolving a merge conflict via github.com

mattallan commented:
Thanks for testing this @Mayisha and bringing this to my attention.
This is actually a bug in WooPayments as well!

> When I click GPay for the virtual product the price is off.

This has sent me down quite the rabbit hole! 😄 The reason why the price is off is because of the way we've written the Subscriptions Synchronisation class and the tldr version is that we rely on the synced product being in the cart to properly mock the trial period and also calculate the initial payment (i.e. prorated or not). This means we can't set the correct price for a synced subscription when preparing the payment request data on page load or variation change.

The only way we can properly calculate the totals is on the `shippingoptionchange` and `shippingaddresschange` payment request events is triggered after the product is in the cart, but problem is that these events don't get triggered for virtual products.

I couldn't find any other event we can useI think the only plausible fix at the moment is to just not display the payment request buttons on the product page when the selected product/variation is a synced virtual product with no upfront payment.

I'll work on this fix and push it up.

Commit:
Refactor how we check for invalid subscription products to support more cases

Commit:
is_product() check doesn't work when inside an ajax request

mattallan commented:
Hey @Mayisha I just pushed up the fix to this problem see [3e44fa4](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2741/commits/3e44fa4900e21bcbc8d7efd8e65d810f13d83336).

This commit updates and renames the recently added `product_has_trial_and_needs_shipping()` function to `is_invalid_subscription_product()` and adds logic for the new case you found.
Usually I would deprecate the `product_has_trial_and_needs_shipping()` function but since it was only added in the last version I'm not sure if it's still necessary. Keen to get your thoughts on that decision though 😅 

Also, to help with understanding this function, I refactored it so that it now lists off all of the invalid cases, rather than trying to work out the inverse of each invalid case.

This is ready for another review! Thanks again

mattallan commented:
Sorry for the runaround but I just found another issue...
When you select an invalid subscription variation and add it to the cart, the PRB get displayed after on page reloads with the invalid variation still selected 😖 

This is because on page load we don't know that a specific variation is selected and so we're showing the button for the whole variable subscription.

https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/fb20403e-d5bd-4485-b17c-a96c2d878776

Mayisha commented:
> Usually I would deprecate the product_has_trial_and_needs_shipping() function but since it was only added in the last version I'm not sure if it's still necessary. Keen to get your thoughts on that decision though

I also think deprecation is not very much required here.

I wanted to confirm one thing for the next review.
> When I click GPay for the virtual product the price is off.

This case I mentioned before for a virtual product with sync on 4th day (price 100), are we supposed to see the PRB or should be hidden because of the sync? :thinking:

mattallan commented:
> This case I mentioned before for a virtual product with sync on 4th day (price 100), are we supposed to see the PRB or should be hidden because of the sync?

This is a complicated case which I tried to document within these [inline comments](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2741/commits/3e44fa4900e21bcbc8d7efd8e65d810f13d83336#diff-e2b9a082adab276b037cd70074b84f278c903bfed3c4b3d77a46aa04bcbfb78aR647-R656) but I'll try and explain it here (hopefully this makes sense 😄)

The short answer is **no**, the Payment request buttons shouldn't be visible when this variation is selected on the product page, however, this variation can be purchased with PRB once it's in the cart from the cart & checkout pages.

A bit confusing... but the reason for this is because while on the product page, we cannot properly calculate the price for the synced virtual product as all of the mock trials and price calculationss are attached to `WC->cart->calculate_totals()` process.
The reason why we can support purchasing physical synced products and not virtual products from the product page is because the Stripe Payment Request APIs trigger the `shippingaddresschange` when the shipping address is selected, which we hook onto and manually call `calculate_totals()` which calculates the proper price of the sync.

---

Sorry for the slow reply, I've been a bit stumped by this issue I found in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2741#issuecomment-1807463674. ~I'm at the stage where I think the only reasonable solution is going back to only showing the PRB if all subscription variations support PRB (which is what we currently have on `develop`).~ I found a way to fix this which means we don't need to do this.

The issue is after you add an invalid variation to your cart using the regular Sign Up button, the PRB get's shown letting you purchase an invalid product with Google Pay and Apple Pay.
The reason for this is because after adding it to the cart, the page reloads with the invalid variation select, and the PRB are shown because at least one variation of the variable subscription supports PRB.

Sorry for the confusing issue!!

~~I will say I think there are benefits to simplifying the logic and only showing the PRB if all variations are supported because I can imagine it might be confusing for customers to see the PRB showing and hiding depending on what variation they are selecting. It could appear as a bug and not immediately obvious that you can't use the PRB buttons for trial products that need shipping~~

Commit:
Fix issues with variations that don't support PRB being selected on page load

mattallan commented:
I've just pushed up [8fae171](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2741/commits/8fae1713e63bcc2a87e975ff6b4ee8850718ee0e) which changes things again 😄 

I've figured out how to fix the above issue, and I think this could've been considered a bug because if you passed variation attributes as GET params to the product URL to select a variation, if that variation doesn't support PRB we would still load them on product page load.

Commit:
Add changelogs

mattallan committed:
Merge branch 'develop' into 2395-further-improvements

Commit:
use correct formatting in changelog entries

Commit:
Add new changelog entry

Commit:
Add the signup fee to the product price for subscription variations as well

Commit:
Allow variation IDs to be added to cart via AJAX when google/apple pay is used

See related PR in WooPayments: https://github.com/Automattic/woocommerce-payments/pull/7702

mattallan commented:
Hey @Mayisha, I'm sorry for the size of this PR and the number of things it's fixing!
The more I pulled on the thread, the more issues I found.

Next time I'll try and be mindful of making my PR focused on one fix at time 😄  


To recap all of the changes/fixes in this PR:
- Renamed `product_has_trial_and_needs_shipping()` to `is_invalid_subscription_product()`
- Refactored `is_invalid_subscription_product( $product )` so that when a variable subscription is passed, it only returns true if all variations don't support using PRB
- Added a new check to `is_invalid_subscription_product()` to fix incorrect PRB totals when purchasing a virtual synced subscription product from the product page
- After adding a variation to the cart or when loading the page with the attributes set in `$_GET` params, don't show the PRBs if the selected variation is not supported.
- Don't require a shipping address when purchasing virtual variable subscription using the PRBs
  - Follow-up issue: signup fees not calculated on virtual variable products (#2755)
- Carry over a recent fix I found from WooPayments https://github.com/Automattic/woocommerce-payments/pull/7702. This mostly fixes issues on stores that have a customized product page where you add variation IDs to the cart

Mayisha reviewed (approved):
Thanks @mattallan for fixing so many issues related to the PRB. :rocket: 
Changes look good with the nicely elaborated comments. All the issues mentioned in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2741#issuecomment-1809511066 are fixed. :tada:
